File: days/src/main/scala/aoc2024s/day4/Part2Grid.scala
package aoc2024s.day4

class Part2Grid(lines: List[String]) {
  val grid: Vector[String] = lines.toVector
  val height: Int = grid.size
  val width: Int = grid(0).length

  def charAt(x: Int, y: Int): Char = {
    grid(y)(x)
  }

  def count(target: String): Long = {
    assert(target.length == 3)
    (for
      x <- 1 until width - 1
      y <- 1 until height - 1
      if charAt(x, y) == target(1)
      if charAt(x - 1, y - 1) == target(0) && charAt(x + 1, y + 1) == target(2)
        || charAt(x - 1, y - 1) == target(2) && charAt(x + 1, y + 1) == target(0)
      if charAt(x + 1, y - 1) == target(0) && charAt(x - 1, y + 1) == target(2)
        || charAt(x + 1, y - 1) == target(2) && charAt(x - 1, y + 1) == target(0)
    yield ()).size
  }
}

File: days/src/main/java/aoc2024j/day4/Part2Grid.java
package aoc2024j.day4;

import utils.CharGrid;

import java.util.List;

public class Part2Grid extends CharGrid {
    private final String target;

    public Part2Grid(List<String> data, String target) {
        assert target.length() == 3;
        super(data, false);
        this.target = target;
    }

    public long count() {
        long counter = 0L;
        for (int y = 1; y < height - 1; y++) {
            for (int x = 1; x < width - 1; x++) {
                counter += countTarget(x, y);
            }
        }
        return counter;
    }

    private long countTarget(int x, int y) {
        if (points[y][x] != target.charAt(1)) {
            return 0L;
        }
        if (diagonal(x, y) && antiDiagonal(x, y)) {
            return 1L;
        }
        return 0L;
    }

    private boolean diagonal(int x, int y) {
        return (points[y - 1][x - 1] == target.charAt(0) && points[y + 1][x + 1] == target.charAt(2))
                || (points[y - 1][x - 1] == target.charAt(2) && points[y + 1][x + 1] == target.charAt(0));
    }

    private boolean antiDiagonal(int x, int y) {
        return (points[y - 1][x + 1] == target.charAt(0) && points[y + 1][x - 1] == target.charAt(2))
                || (points[y - 1][x + 1] == target.charAt(2) && points[y + 1][x - 1] == target.charAt(0));
    }
}
